클라이밍장 조회, 추가 API 구현

wonyangs opened the pull request:
클라이밍장 단건, 다건 조회 및 추가 API를 구현하였습니다.

## 클라이밍장 리스트 조회 (다건 조회)

### 예시
- request: `GET /gym`
- response: 200 OK
```json
[
    {
        "id": 1,
        "gymName": "hanyuClimbingGym",
        "thumbnailUrl": "example.png"
    },
    {
        "id": 2,
        "gymName": "hanyuClimbingGym 논현점",
        "thumbnailUrl": "AwesomeThumbnail.png"
    },
    {
        "id": 3,
        "gymName": "hanyuClimbingGym 강남점",
        "thumbnailUrl": "AwesomeThumbnail2.png"
    }
]
```

## 클라이밍장 정보 조회 (단건 조회)

- API 설계에는 클라이밍장에 포함되어 있는 Sector 정보를 전달
```
{
    "mapImageUrl": string,
    "difficulties": string[],
    "sectors": string[]
}
```
- querydsl 적용 예정이기 때문에 우선 아래와 같이 구현한 상태로 병합하기로 결정
  - 추후 querydsl 적용 시 함께 수정 예정

### 예시
- request: `GET /gym/{gymId}`
- response: 200 OK
```json
{
    "id": 3,
    "gymName": "hanyuClimbingGym 강남점",
    "thumbnailUrl": "AwesomeThumbnail2.png",
    "mapImageUrl": "map2.png",
    "difficulties": [
        {
            "name": "파랑",
            "level": 1
        },
        {
            "name": "빨강",
            "level": 2
        },
        {
            "name": "보라",
            "level": 3
        }
    ]
}
```

## 클라이밍장 생성

- 기존 설계한 API 명세에는 이미지 Url을 받도록 되어 있음
  - 현재 브랜치에서는 기존 명세대로 구현되어 있음
- 이미지 업로드를 서버에서 처리하기로 결정됨
  - 클라이언트에서 이미지를 다루는 것이 보안상 위험하다고 판단
- 추후에 이미지 업로드 파트를 분리할 예정이기 때문에 Url 인자가 수정될 예정
  - thumbnail, mapImage를 이미지 파일로 받아 업로드
- 응답 본문을 위해 `ClimbingGymCreateResponseDto`을 제작

### 예시
- request: `POST /gym`
```json
{
    "gymName": "hanyuClimbingGym 강남점",
    "thumbnailUrl": "AwesomeThumbnail2.png",
    "mapImageUrl": "map2.png",
    "difficulties": [
		{
            "name": "파랑",
            "level": 1
        },
        {
            "name": "빨강",
            "level": 2
        },
        {
            "name": "보라",
            "level": 3
        }
	]
}

```
- response: 201 CREATED
```json
{
    "id": 3,
    "gymName": "hanyuClimbingGym 강남점",
    "thumbnailUrl": "AwesomeThumbnail2.png",
    "mapImageUrl": "map2.png",
    "difficulties": [
        {
            "name": "파랑",
            "level": 1
        },
        {
            "name": "빨강",
            "level": 2
        },
        {
            "name": "보라",
            "level": 3
        }
    ]
}
```

## 기타 수정 사항

- 테스트명을 한글로 수정
- 코드 포맷이 google style로 되어있지 않던 부분 수정
- 패키지 유지를 위한 임시 파일 삭제

wonyangs committed:
FLASH-128 chore: 테스트명을 한국어로 수정

wonyangs committed:
FLASH-128 feat: 모든 클라이밍장 리스트를 반환하는 서비스 메서드 구현
- Repository에 findAll을 추가

wonyangs committed:
FLASH-128 fix: save 메서드의 반환값을 Long으로 수정

wonyangs committed:
FLASH-128 test: 다건 조회 테스트 추가

wonyangs committed:
FLASH-128 chore: 코드 포맷 수정

wonyangs committed:
FLASH-128 chore: 임시 파일 삭제

wonyangs committed:
FLASH-128 feat: save 메서드의 반환값을 entity로 수정

- 응답 body에 entity 내용을 담기 위해 수정

wonyangs committed:
FLASH-128 feat: 클라이밍장 조회, 생성 API 구현

- 응답에 담긴 id 키를 gymId에서 id로 수정
- 클라이밍장 생성 시 응답에 담을 정보를 위한 dto 구현

wonyangs committed:
FLASH-128 Merge branch 'integration' into feature/FLASH-128-climbing-gym-view-api

wonyangs committed:
FLASH-128 fix: 테스트코드에 save 메서드 반환값 변경이 반영 안된 부분 수정

ChoiWonYu commented on src/test/java/com/first/flash/climbing/gym/application/ClimbingGymServiceTest.java:
```diff
@@ -6,18 +6,19 @@
 import org.springframework.boot.test.context.SpringBootTest;
 
 @SpringBootTest
 class ClimbingGymServiceTest {
```
ServiceTest에 대해 유닛 테스트를 하기 위해선 제가 한 방식이나 mock 객체를 이용하는 것이 좋을 것 같습니다!
다음 스프린트 때에는 테스트의 범위, 대상, 방법 등에 대한 명확한 기준을 세우는 게 좋을 것 같네요!

wonyangs replied:
ServiceTest 개선한다고 해놓고 빠트렸네요😅
SectorServiceTest와 같은 방식으로 FakeRepository를 만들어 개선하였습니다.

다음에 테스트 방법 등을 명확히 세우면 좋을 것 같습니다!

ChoiWonYu commented on src/main/java/com/first/flash/climbing/gym/ui/ClimbingGymController.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    @GetMapping("/{gymId}")
+    public ClimbingGym getGymDetails(@PathVariable Long gymId) {
```
여기 final이 빠져 있네요!

ChoiWonYu commented on src/test/java/com/first/flash/climbing/gym/application/ClimbingGymServiceTest.java:
```diff
@@ -59,4 +60,21 @@ class ClimbingGymServiceTest {
+        List<ClimbingGymResponseDto> gyms = climbingGymService.findAllClimbingGyms();
+
+        // then
+        assertThat(gyms).isNotEmpty();
```
여러 번의 단언을 하는 경우 모든 단언을 테스트하기 위해서 assertSoftly를 사용하는 것이 좋아보입니다!

wonyangs replied:
```java
assertSoftly(softly -> {
    softly.assertThat(gyms).isNotEmpty();
    softly.assertThat(gyms.size()).isEqualTo(2);
});
```
assertSoftly를 사용하여 위와 같이 수정하였습니다!

ChoiWonYu reviewed (commented):
전체적으로 깔끔하게 잘 하신 것 같습니다! 고생하셨습니다🙌

wonyangs committed:
FLASH-128 refactor: 여러 단언을 테스트하기 위해 assertSoftly로 수정

wonyangs committed:
FLASH-128 chore: 함수 인자에 final이 빠져있던 부분 수정

wonyangs committed:
FLASH-128 refactor: ServiceTest 동작 방식 수정

- 기존 @SpringbootTest 방식 개선
- FakeClimbingGymRepository 구현
- ClimbingGymService에 주입하여 구현

wonyangs committed:
FLASH-128 chore: 테스트코드에 접근제한자 추가
